lib/puppet/pops/issues.rb

```diff
@@ -490,6 +490,14 @@ def self.hard_issue(issue_code, *args, &block)
     _("Unacceptable location. The name '%{name}' is unacceptable in file '%{file}'") % { name: name, file: file }
   end
 
+  ILLEGAL_TOP_CONSTRUCT_LOCATION = issue :ILLEGAL_TOP_CONSTRUCT_LOCATION, :name do
+    if name == :no_name
+      _("Unacceptable location. Unacceptable top level construct")
+    else
+      _("Unacceptable location. Unacceptable top level construct '%{name}''") % { name: name }
```

Comment:
You have two single quotes at the end of the message here.

Reply:
I wonder about the phrasing of these messages - isn't it rather something like "Unacceptable as a top level expression for logic in this location"? The current message to me sounds like it is telling med that I should have the construct someplace else (i.e. the location is unacceptable and the construct would work if only it was in a different location - and that is probably not true).

Could also include the type of the element to blame: for example :   `_("This %{value} is unacceptable as a top level construct in this location") % { value: label(semantic) }`
Which would output something like `"This If Expression is unacceptable..."` (See `model_label_provider.rb` to get a feel for what it will output for the respective reported AST elements).
(That is if you want to include the kind of expression that is not allowed).

I think I would have split the issue into two; one for named elements (`ILLEGAL_NAMED_TOP_CONSTRUCT_LOCATION`) and one for unnamed. That way, all `semantic` given to the named variant will be `NamedDefinitions` and you can assume `semantic.name` will return the name.

Reply:
You could do this:
```
name = semantic.is_a?(Puppet::Pops::Model::NamedDefinition) ? semantic.name : nil
if name.nil?
  _("This %{value} is unacceptable as a top level construct in this location") % { value: label(semantic) }
else
  _("The %{value} '%{name}' is unacceptable as a top level construct in this location") % { name: name, value: label(semantic) }
end
```
To (for example) produce `The Function Definition 'foobar' is unacceptable ...`

You would then not need to have `:name` as a parameter to the issue, and there is no need in checking if something responds to `:name` when making the call.

These are all suggestions for improvements should you like to include the name and kind of element.